crypto/common/mod.go

```diff
@@ -0,0 +1,110 @@
+// Package common implements interfaces to support multiple algorithms. A public
+// key factory and a signature factory are available. The supported algorithms
+// are the followings:
+// - BLS
+package common
+
+import (
+	"reflect"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"go.dedis.ch/fabric/crypto"
+	"go.dedis.ch/fabric/crypto/bls"
+	"go.dedis.ch/fabric/encoding"
+	"golang.org/x/xerrors"
+)
+
+// PublicKeyFactory is a public key factory for commonly known algorithms.
+//
+// - implements crypto.PublicKeyFactory
+type PublicKeyFactory struct {
+	encoder   encoding.ProtoMarshaler
+	factories map[reflect.Type]crypto.PublicKeyFactory
+}
+
+// NewPublicKeyFactory returns a new instance of the common public key factory.
+func NewPublicKeyFactory() PublicKeyFactory {
+	factory := PublicKeyFactory{
+		encoder:   encoding.NewProtoEncoder(),
+		factories: make(map[reflect.Type]crypto.PublicKeyFactory),
+	}
+
+	factory.Register((*bls.PublicKeyProto)(nil), bls.NewPublicKeyFactory())
+
+	return factory
+}
+
+// Register binds a protobuf message type to a public key factory. If a key
+// already exists, it will override it.
+func (f PublicKeyFactory) Register(msg proto.Message, factory crypto.PublicKeyFactory) {
```

Comment:
It seems more natural and user-friendly because you want to register a type of message to be handled by a factory. Also we can switch to something else later if reflect is problematic.